src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -0,0 +1,709 @@
+    let max = dataset?.getFeatureData(featureName)?.max || 0;
+
+    // Special case for time feature, which isn't in the dataset
+    if (featureName === TIME_FEATURE.name) {
+      min = 0;
+      max = dataset?.numberOfFrames || 0;
+    }
+
+    if (dataset && dataset.isFeatureCategorical(featureName)) {
+      // Add extra padding for categories so they're nicely centered
+      min -= 0.5;
+      max += 0.5;
+    } else {
+      // Add a little padding to the min/max so points aren't cut off by the edge of the plot.
+      // (ideally this would be a pixel padding, but plotly doesn't support that.)
+      min -= (max - min) / 10;
+      max += (max - min) / 10;
+    }
+    scatterPlotAxis.range = [min, max];
+
+    // TODO: Show categories as box and whisker plots instead of scatterplot?
+    // TODO: Add special handling for integer features once implemented, so their histograms use reasonable
+    // bin sizes to prevent jumping.
+
+    if (dataset && dataset.isFeatureCategorical(featureName)) {
+      // Create custom tick marks for the categories
+      const categories = dataset.getFeatureCategories(featureName) || [];
+      scatterPlotAxis = {
+        ...scatterPlotAxis,
+        tickmode: "array",
+        tick0: "0", // start at 0
+        dtick: "1", // tick increment is 1
+        tickvals: [...Array(categories.length).keys()], // map from category index to category label
+        ticktext: categories,
+        zeroline: false,
+      };
+      // Enforce bins on histogram traces for categorical features. This prevents a bug where the histograms
+      // would suddenly change width if a category wasn't present in the given data range.
+      newHistogramTrace.xbins = { start: min, end: max, size: (max - min) / categories.length };
+      // @ts-ignore. TODO: Update once the plotly types are updated.
+      newHistogramTrace.ybins = { start: min, end: max, size: (max - min) / categories.length };
+    }
+    return { scatterPlotAxis, histogramAxis, histogramTrace: newHistogramTrace };
+  };
+
+  /**
+   * VERY roughly estimate the max width in pixels needed for a categorical feature.
+   */
+  const estimateTextWidthPxForCategories = (featureName: string): number => {
+    if (featureName === null || !dataset?.isFeatureCategorical(featureName)) {
+      return 0;
+    }
+    const categories = dataset.getFeatureCategories(featureName) || [];
+    return (
+      categories.reduce((_prev: any, val: string, acc: number) => {
+        return Math.max(val.length, acc);
+      }, 0) * 8
+    );
+  };
+
+  //////////////////////////////////
+  // Plot Rendering
+  //////////////////////////////////
+
+  const plotDependencies = [
+    dataset,
+    xAxisFeatureName,
+    yAxisFeatureName,
+    rangeType,
+    currentFrame,
+    selectedTrack,
+    props.isVisible,
+    isPlaying,
+    plotDivRef.current,
+  ];
+
+  const renderPlot = useCallback((forceRelayout: boolean = false) => {
```

Comment:
Nit: Not really crucial for this PR, but this `renderPlot` is pretty long.  I wonder if it could be split up into smaller function pieces, like the one that builds the layout, the one that builds the traces, etc?  I would almost consider splitting this plotly handling code into its own class and farming some of these details out into that class.

Reply:
I made some improvements to this in the next PR, and I'm slowly moving utility functions out into separate files.

Reply:
Good question, actually. I'm realizing that this doesn't need to be in one, since it's not being passed as a prop anywhere. Fixed!